### PR TITLE
Fix GTP checkpoint load failing across different GTP degrees

### LINKED ISSUE #6501: [feat] GTP+DCP
GTP+DCP(Distributed Checkpoint) has been supported in the GTP init MR #4967 

This issue will keep tracking following-up bug-fixes / perf optimizations for GTP+DCP related stuffs.
- fix GTP+DCP for GDP moudle (#6503 )
- fix GTP+DCP when parallel_state is not initialized (#6775 )
- fix the case when some params is sharded across GTP while been duplicated across TP group (e.g., the latent gemm) (#6781 )
- GTP+DCP+fp8-param save->load is not bitwise matching 
        - MXFP8 (bytes + scale_A) → dequantize → BF16 → save → load → quantize → MXFP8 (bytes + scale_B)
               - Where scale_A (from fp32 master w) ≠ scale_B
               - quantize(dequantize(mfxp8) is a lossy round-trip, not identical
        - solution: during loading: cast from fp32 main params into mxfp8 params instead (related MR: #6666 )
- Fix GTP checkpoint load failing across different GTP degrees (#7293)

### PR DESCRIPTION
**Pull request by fanshiqing**
- [x] I, the PR author, have personally reviewed every line of this PR.

## Summary
A checkpoint saved with GTP (Generalized Tensor Parallelism) alignment padding at one `gtp_remat_size` could fail to load at a different `gtp_remat_size`, or when loaded by a non-GTP process — the checkpoint's declared shape (padded) didn't match the loader's expected shape (possibly unpadded or padded differently), and `allow_shape_mismatch` was never set to tolerate the difference.

Here are two typical cases this MR should fix:

### Example-1: GTP-save-with-padding -> GTP-load-w/o-padding
image

### Example-2: GTP-save --> GTP load but with different padding requirement
image

## Coverage matrix

| Save \ Load | Plain 3D (TP×PP×DP, no GTP) | GTP (any `gtp_remat_size`) |
|---|---|---|
| **Plain 3D (no GTP)** | n/a — no padding involved, unaffected | Fixed — load-side padding recognized (`declared == dim0_unpadded`, e.g. non-tile-aligned hidden dims) |
| **GTP (padded)** | Fixed — the originally reported gap; a GTP-saved checkpoint loaded by a GTP-unaware process now recognizes the padding | Fixed — the originally reported bug; save/load at different GTP degrees now reconciles correctly |

All "Fixed" cells are validated for plain TP×DP. **Pipeline parallelism (PP > 1) is a known gap**, not covered by this fix — see below.

## Fix

Add `infer_gtp_allow_shape_mismatch()` (`megatron/core/utils.py`), called right before `dist_checkpointing.load()`. It reads the checkpoint's real on-disk shape and sets `allow_shape_mismatch` per tensor — but only when the declared-vs-expected difference is actually explainable as GTP padding (bounded by the recipe's alignment size), not for every mismatch.

`pad_for_alignment` (the recipe's alignment tile size) is resolved via a new `megatron/core/tensor_parallel/gtp_recipe.py` — a small, zero-dependency module so it can be computed during checkpoint load without importing TE-dependent GTP modules.

## Testing

- **Real end-to-end functional verification**: `test_cross_gtp_degree_save_load_roundtrip` does a genuine `dist_checkpointing.save()`/`load()` roundtrip for the originally-reported case — GTP4-padded save (`dim0=192` → 256) loaded by GTP2 (no padding needed, `192 % 64 == 0`). The load-side buffer is pre-scribbled with a sentinel value before loading (so a no-op or partial load can't accidentally pass), and the result is checked bit-exact (`rtol=0, atol=0`) against an independently-constructed golden tensor. This confirms the fix doesn't just flip a boolean — the actual loaded weight data is correct.
- Additional unit tests pin the padding-explainability bound directly: legitimate padding, an unrelated/incompatible shape (must stay rejected), a GTP-saved checkpoint loaded by a non-GTP process, a pre-set `allow_shape_mismatch` from an unrelated reason (must not be clobbered), and the bf16 (`pad_for_alignment=1`) case (must not vacuously accept any size).
- Full `tests/unit_tests/generalized_tensor_parallel/` suite (all 18 files, ~200+ cases) passes cleanly, including this one.

## Known limitations (not in the scope of this MR)
- `SwiGLU linear_fc1.weight + GTP` is unreconciled by this fix
     - and separately, there's an open question about whether the GTP+SwiGLU factory split is even correct pre-existing, worth its own dedicated investigation.
- The padding-explainability check has no upper bound on accepted shape growth beyond
  alignment/exact-match.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

File: megatron/core/utils.py [hunk 1]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+        pads             declared > expected declared != expected (different pad amounts)
+
+    ``pad_for_alignment`` must come from the caller (the training recipe), not ``GTP_CONFIG`` --
+    that global is only set when GTP is active in *this* process.
+    """
+    from megatron.core.dist_checkpointing.dict_utils import nested_values
+
+    sharded_tensors = [v for v in nested_values(sharded_state_dict) if isinstance(v, ShardedTensor)]
+    if not sharded_tensors:
+        return
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply megatron/core/utils.py:1185]: `SwiGLU linear_fc1.weight + GTP` is unreconciled by this fix.
and separately, there's an open question about whether the GTP+SwiGLU factory split is even correct pre-existing, worth its own dedicated investigation.

Let's keep it out of this MR and should be resolved by some other following up MRs.

File: megatron/core/utils.py [hunk 2]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+        dim0_unpadded = expected0 - int(getattr(sh_ten, "gtp_pad_length", 0))
+        # Valid padding: either an unpadded save (declared0 == dim0_unpadded exactly), or an
+        # alignment-padded one (declared0 a multiple of pad_for_alignment). At
+        # pad_for_alignment == 1 (bf16) every integer is "a multiple", so that disjunct is gated
+        # off there -- otherwise any declared0 >= expected0 would pass as "padding".
+        is_valid_padding = declared0 == dim0_unpadded or (
+            pad_for_alignment > 1 and declared0 % pad_for_alignment == 0
+        )
+        # Padding only ever adds rows, so also require declared0 >= dim0_unpadded.
+        sh_ten.allow_shape_mismatch = declared0 >= dim0_unpadded and is_valid_padding
```
[the remark this answers is not in the document]
Comment (fanshiqing) [reply megatron/core/utils.py]: I have thinked about it and there's no cheap, reliably-correct bound to add without real additional plumbing.
So let's keep it out of the  scope for the current MR.

**Review by kvareddy — APPROVED** [review]

File: megatron/core/utils.py [hunk 3]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+        # The GTP-padded axis is dim0 of the WEIGHT, not necessarily index 0 of global_shape --
+        # e.g. a PP-layer axis prepended via prepend_offsets/sharded_offsets shifts it to
+        # prepend_axis_num. The checkpoint-side ShardedTensor is metadata-only (its own
+        # prepend_axis_num is meaningless, always 0), so index both shapes by the live tensor's.
+        axis0 = sh_ten.prepend_axis_num
+        if axis0 >= len(ckpt_shape):
+            continue  # axis layout mismatch -- not a shape this function understands
+        declared0 = int(ckpt_shape[axis0])
+        expected0 = int(sh_ten.global_shape[axis0])
+        if declared0 == expected0:
```
Comment (deepakn94) [inline megatron/core/utils.py]: `expected` and `declared` are a bit confusing...can we make one of these `ckpt_...` to signify the one that's coming from the checkpoint?

File: megatron/core/utils.py [hunk 4]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+            continue  # no difference -- nothing to fix
+        dim0_unpadded = expected0 - int(getattr(sh_ten, "gtp_pad_length", 0))
+        # Valid padding: either an unpadded save (declared0 == dim0_unpadded exactly), or an
+        # alignment-padded one (declared0 a multiple of pad_for_alignment). At
+        # pad_for_alignment == 1 (bf16) every integer is "a multiple", so that disjunct is gated
+        # off there -- otherwise any declared0 >= expected0 would pass as "padding".
+        is_valid_padding = declared0 == dim0_unpadded or (
```
Comment (deepakn94) [inline megatron/core/utils.py]: Can't we just compute what the unpadded sizes are in GPU memory and in the checkpoint, and then check for equality?

Comment (fanshiqing) [reply megatron/core/utils.py]: **No we can't make it today,** because for the `checkpoint side` the `.metadata` did not record with `gtp_remat_size` produced the padding during saving, and only the final padded shape is saved.

Only with the saving time's `gtp_remat_size` then we can calculate the `checkpoint side`'s unpadded size.

File: megatron/core/utils.py (hunk 3 above)
Comment (fanshiqing) [reply megatron/core/utils.py]: good point, will fix

**Review by maanug-nv — APPROVED** [review]

**Review by Mellonta — APPROVED** [review]

File: megatron/core/utils.py [hunk 5]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+    if fp8_recipe == "mxfp8":
+        return 32
+    if fp8:
+        return 16
+    # No MXFP8/NVFP4 tile-size requirement in this recipe -- pad only to the minimum
+    # gtp_remat_size needed for even AG/RS sharding, not a fixed quantization tile size.
+    return 1
+
+
+def infer_gtp_allow_shape_mismatch(sharded_state_dict, checkpoint_dir, pad_for_alignment):
```
Comment (Phlip79) [inline megatron/core/utils.py]: Can this be changed to `check_gtp_shape_mismatch` or something without "infer" which is commonly used to reference "inference".

File: megatron/core/utils.py [hunk 6]
```diff
@@ -1136,9 +1136,90 @@ def make_tp_sharded_tensor_for_checkpoint(
...
+    sharded_tensors = [v for v in nested_values(sharded_state_dict) if isinstance(v, ShardedTensor)]
+    if not sharded_tensors:
+        return
+
+    try:
+        from megatron.core.dist_checkpointing.serialization import load_tensors_metadata
+
+        checkpoint_metadata = load_tensors_metadata(str(checkpoint_dir))
+    except Exception:  # noqa: BLE001
+        return  # can't read metadata -> leave flags as-is, don't block the load on this check
```
Comment (Phlip79) [inline megatron/core/utils.py]: When would this actually happen? Should there be a warning?

File: megatron/training/checkpointing.py [hunk 7]
```diff
@@ -1971,6 +1974,16 @@ def _load_global_dist_base_checkpoint(
         )
     if checkpointing_context is not None:
         checkpointing_context['load_strategy'] = load_strategy
+    from megatron.core.utils import infer_gtp_allow_shape_mismatch, resolve_gtp_pad_for_alignment
```
Comment (Phlip79) [inline megatron/training/checkpointing.py]: Can you move the import to the top of the file?

File: megatron/core/utils.py (hunk 6 above)
Comment (fanshiqing) [reply megatron/core/utils.py]: The exceptions happens when the checkpoint isn't readable yet or is in a bad state.
It's a good idea we should at least add some warning here instead of return siliently. 
will do

File: megatron/training/checkpointing.py (hunk 7 above)
Comment (fanshiqing) [reply megatron/training/checkpointing.py]: Sure, it should be

File: megatron/core/utils.py (hunk 5 above)
Comment (fanshiqing) [reply megatron/core/utils.py]: okay, I will rename it to `grant_shape_mismatch_for_gtp_padding` later

**Review by Phlip79 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]
